proposals/0061-fair-congestion-control-with-intra-block-exponential-local-base-fee.md

```diff
@@ -0,0 +1,127 @@
+---
+simd: '0061'
+title: Fair congestion control with intra-block exponential local base fee
+authors:
+  - Ryo Onodera (Solana Labs)
+category: Standard/Meta
+type: Core
+status: Draft
+created: 2023-07-05
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This feature is a fair congestion-control mechanism in the form of an extension
+to local fee markets while leaving their locality of transaction fee dynamics
+intact.
+
+To that end, it introduces a exponentially-scaled dynamic base fees to
+individual local fee markets.  It also attains very short feedback loop of
+intra block frequency to maintain full efficacy of Solana's peculiar execution
+model compared to other blockchains: multi-threaded and low latency.
+
+This is realized by means of some incentive tweaks to combat against the
+obvious base fee manipulation with such short intervals
+
+## Motivation
+
+- Write lock cu limit is bad (bot can lock out at the very first of block for
+  the entire duration of whole blocktime (400ms)
+- Increased Defi activities around any volatile financial markets could starve
+  payment transactions for extended time
+- Inter-block and linear Voluntary fee escalation with vanilla fee market
+  auction can't guarantee the scheduling deadline of casual payment txes (which
+  needs 99.99% sub-second confirmation at very minimum).
+
+## Alternatives Considered
+
+Related proposals:
+
+(TODO: add any relation of this to them)
+
+dynamic base fees
+https://github.com/solana-foundation/solana-improvement-documents/pull/4
+
+program rebatable account write fees:
+https://github.com/solana-foundation/solana-improvement-documents/pull/16
+
+asynchronous program execution:
+https://github.com/solana-foundation/solana-improvement-documents/pull/45
+
+increase prioritization fee:
+https://github.com/solana-foundation/solana-improvement-documents/pull/50
+
+bankless
+https://github.com/solana-foundation/solana-improvement-documents/pull/5
+
+## New Terminology
+
+Is there any new terminology introduced with this proposal?
+
+casual tx:
+fairness:
+block fullness in terms of number of actively-execution threads
+
+## Detailed Design
+
+(i jotted this down in 10min before going to bed! pardon for being so random
+writings...)
+
+to *determiniscally* define active thread count (`TC_a`), additionally record
+transaction termination events into poh stream.
```

Comment:
How does this give us thread count? PoH recording is single-threaded.

Reply:
this opens up for a special type of MEV where the leader can artificially inflate the base fee, e.g. mango's validator could increase the base fee for zeta traders.

Reply:
hey, thanks for jumping in.

like this?

serialized new ledger entries:
1. SanitizedTransactionA
2. TerminationMarkerForTransactionA
3. SanitizedTransactionB
4. SanitizedTransactionC
5. TerminationMarkerforTransactionB
6. TerminationMarkerForTranactiionC



```
 ..1: TC_a: 0
1..2: TC_a: 1
2..3: TC_a: 0
3..4: TC_a: 1
4..5: TC_a: 2
5..6: TC_a: 1
6.. : TC_a: 0
```

this proposal assumes a single threaded banking scheduler, not thread-local-MI.

so, banking threads newly start to record transaction terminations via the scheduler thread. and single threaded poh can be able to linearize these events without ambiguity and replay stage can reconstruct `TC_a` just given list of these ledger entries.

Reply:
It seems as a greedy leader I can lie about the termination in order to maximize parallelism, in order to raise fees. Seems like it'd be easy to just keep a cache of the terminations and just record them right at the end of the slot.

Even with conflicts, could just record & (remove from cache) cached tx-terminations that are in conflict.

Reply:
yeah, so, this proposal strongly encodes disincentivation of parallelism manipulation. as you pointed out  base fee is indeed raised, but throughput is forcibly limited. the rate limit is adjusted to that total transaction fee after manipution would never greater than scheduled normally.

in this proposal, cu & active thread are both available for determisnic replaying.